PartDesign: Auto switch label to edit after adding new part

ashimabu opened the pull request:
After adding a new body, the tree will auto switch the body label to edit mode so a user can type a name.

There are several approaches that could work here but tracking state on the DocumentObject and having the tree react was cleanest.

NEW BEHAVIOR
![relabelNewBehavior](https://github.com/user-attachments/assets/edf934af-0f40-4cde-8992-d8b9a912b27a)

OLD BEHAVIOR
![relabelOldBehavior](https://github.com/user-attachments/assets/f4852c7b-11ad-4bf5-b423-9f5edc9c19f6)

kadet1090 commented on src/Gui/Tree.cpp:
```diff
@@ -3193,6 +3195,28 @@ void TreeWidget::onUpdateStatus()
 
     updateGeometries();
+
+    // offer a relabel for new bodies
+    for (auto& v : localChangedObjects) {
+        auto obj = v.first;
+
+        auto iter = ObjectTable.find(obj);
+        if (iter == ObjectTable.end()) {
+            continue;
+        }
+
+        if (iter->second.empty()) {
+            continue;
+        }
+
+        auto dObj = iter->first;
+        if (dObj->shouldOfferRelabel()) {
+            auto& data = *iter->second.begin();
+            editItem(data->rootItem);
+            dObj->setOfferedRelabel();
+        }
+    }
+
```
This should extracted to separate private method probably.

ashimabu replied:
Can definitely do that, however I tried to match the style of the existing method. It would be a strange method with an argument of type std::unordered_map<App::DocumentObject*,std::bitset<32>>

ashimabu replied:
I ended up abstracting this into a fairly compact private method. Thanks for the suggestion.

kadet1090 reviewed (commented):
I get why this could be useful but I feel that it would split community in half. For me this behavour would be very frustrating.

I suggest adding user parameter for this and exposing it in preferences. I am not sure about default here - personally I'd keep the current behaviour as I cannot recall such feature from any other software or this kind

@FreeCAD/design-working-group what do you think?

maxwxyz commented:
I would also keep the current default.

benj5378 commented:
I generally want to avoid arguments regarding what other software do - it itself doesn't mean it is better or worse. Though we can argue _why_ something other software do is better/worse.

If this behaviour is merged, either by default or through a parameter, then I think it should apply to all objects, if it shouldn't cause inconsistency. However, how will this work if the task panel also needs to be opened upon adding? Some objects are also relevant to name (bodies) while others are likely not (pad, array, TechDraw page).

I'm not sure if I am for or against... though i have a hard time seeing why one behaviour is better than the other (when taking inconsistency with other added objects into account).

benj5378 commented:
> Relying on what other already learned, not repeating their mistakes

I'm just hesitant to assume, that the reason other software doesn't have some feature, is because they already tested it with negative results.

kadet1090 commented:
> > Relying on what other already learned, not repeating their mistakes
> 
> I'm just hesitant to assume, that the reason other software doesn't have some feature, is because they already tested it with negative results.

Agreed, second point is the more important for that specific case, first one is more applicable to how things are done not what features are available. If behaviour is not common then users are not used to it and it will feel strange because FreeCAD will be the outlier. 

That does not of course necessarily mean that it is bad idea but looking at other software is always important clue and providing consistency not only in FreeCAD itself but across whole product category is something that is very important for good UX.

ashimabu commented:
Thanks for the discussion, all. I agree that if this is merged, it should be an opt-in behavior. It would likely fit well in some workflows but would be annoying in others. Very much up to a user's preference. 

I've wrapped it behind a user preference which **defaults to false**. It's in the Selection section because most other Tree related preferences live there. 

![image](https://github.com/user-attachments/assets/23fe2f68-5c94-4a78-85cd-17975315e72e)

@benj5378 it would definitely be nice if this was consistent for all top level tree objects but implementation gets very tricky as you mentioned. It would take a bit of thinking for me to get that working (if possible at all).

furgo16 commented on src/Gui/PreferencePages/DlgSettingsSelection.ui:
```diff
@@ -1,296 +1,309 @@
+			<item row="8" column="0" colspan="2">
+				<widget class="Gui::PrefCheckBox" name="checkBoxTreeAutoRelabel">
+					<property name="text">
+						<string>Auto relabel bodies after they are created</string>
```
Thanks for your work.
1. Could we add an explanatory tooltip as well? What "relabel" means in this context might be obvious to those involved in this PR, but it will be a new concept that needs explanation for users (and translators!).
2. The .ui file has a lot of whitespace changes. The next time someone edits the file with Qt Designer, it will probably undo them. Is it worth doing these whitespace changes?

ashimabu replied:
Thanks for the input. Tool tip added and whitespacing reverted.

![image](https://github.com/user-attachments/assets/caf8f020-4b88-4e72-a074-85842f84a841)

maxwxyz commented:
Is the selection tab the ideal place for this preference? If it only affects PD bodies, it might be better suited in the Part/PD preferences. Regarding the description, maybe it clearer if it is labeled: "Edit body label after creation" instead of "Auto relabel bodies" but I'll leave that to @sifabiancic

@FreeCAD/design-working-group FYI

ashimabu commented:
I updated the code to generalize to all top level tree objects, not just PartDesign. As others have pointed out, it better if the auto relabel feature is more universal.
![relabelNewest](https://github.com/user-attachments/assets/3ff0b33c-cebb-4a69-ad34-b42dfe18a050)

I've also moved the user preference to the Document section as it's a more logical place a user would expect it.
![image](https://github.com/user-attachments/assets/5f9a1617-27a4-45b8-835d-959a529f04c0)

adrianinsaval commented:
This seems like an extremely specific obscure usecase, should we really be adding stuff that only one or two people will use in a very specific situation? IMO this is just clutter if it's going to be limited to bodies and off by default.

> It feels like one of these things which should be use everywhere or nowhere

+1

In any case I think a reasonable compromise could be to make any newly created object selected in the tree so that you can directly hit f2 to rename after creation.

adrianinsaval commented:
I suggest changing the checkbox text "Edit object label after creation", auto relabel is confusing, it sounds like the new label will be automatically generated rather than making the user set it.

adrianinsaval commented:
@FreeCAD/design-working-group would appreciate your input for the checkbox text and also if this should be on or off by default now that it is global and not just for PD

hyarion commented on src/App/DocumentObject.h:
```diff
@@ -75,6 +75,7 @@ enum ObjectStatus
     RecomputeExtension = 19,        // mark the object to recompute its extensions
     TouchOnColorChange = 20,        // inform view provider touch object on color change
     Freeze = 21,                    // do not recompute ever
+    OfferRelabel = 22               // indicates that the object is new and should auto relabel
```
Isn't there a way to implement this `Gui` feature without touching `App`?
For example, can't `New = 2` be used instead to check if the object is new and read out the preference to see if it should offer the user to change the label on the document object?

ashimabu replied:
Tracking state via the DocumentObject's status bits seemed cleanest. Though, I agree it would be ideal to keep this logic contained to the Tree. I wanted to use the New = 2 bit but it's actually only ever set to true and never purged. Readapting this bit would probably require even more changes in App. 

The other option is to keep some map<DocumentObj, bool> on the TreeWidget to track new objects' relabel status. If it's preferred, I can pursue that option.

hyarion replied:
> I wanted to use the New = 2 bit but it's actually only ever set to true and never purged. Readapting this bit would probably require even more changes in App.

What happens if you purge it, do you find any regressions? I don't see it tested anywhere, only set.

ashimabu replied:
I didn't try purging it but it likely wouldn't cause too much issue.

I worked with your suggestion and reverted all the App changes. This functionality is 100% possible to implement in just the TreeWidget. Please take a look and let me know what you think.

Also removed those trailing whitespaces you found.

hyarion commented on src/Gui/Tree.cpp:
```diff
@@ -3061,9 +3066,14 @@ void TreeWidget::onUpdateStatus()
             if (docItem->ObjectMap.count(obj))
                 continue;
-            auto vpd = Base::freecad_dynamic_cast<ViewProviderDocumentObject>(gdoc->getViewProvider(obj));
+            auto vpd = Base::freecad_dynamic_cast<ViewProviderDocumentObject>(gdoc->getViewProvider(obj));            
```
Trailing whitespaces
```suggestion
            auto vpd = Base::freecad_dynamic_cast<ViewProviderDocumentObject>(gdoc->getViewProvider(obj));
```

hyarion commented on src/Gui/TreeParams.cpp:
```diff
@@ -179,7 +182,7 @@ class TreeParamsP: public ParameterGrp::ObserverType {
             return;
         it->second(this);
-
+        
```
Trailing whitespaces
```suggestion

```

maxwxyz commented:
@ashimabu ping for feedback

ashimabu commented:
> I suggest changing the checkbox text "Edit object label after creation", auto relabel is confusing, it sounds like the new label will be automatically generated rather than making the user set it.

Great suggestion. Updated to "Edit object label after creation"

ashimabu commented:
> @ashimabu ping for feedback

Cleaned up a label and contained all the changes to the TreeWidget per some suggestions above. I think this is in a cleaner (and maybe mergeable) state.

kadet1090 reviewed (commented):
Thanks for changes! I think that we are getting quite close to the point where it could be merged. 

Can you explain how the mechanism works and what is the use of `RelabelQueue` and why it is needed? I have some troubles understanding that part - but I don't know this part of code too well so that might be why.

ashimabu commented on src/Gui/Tree.h:
```diff
@@ -226,6 +226,7 @@ private Q_SLOTS:
 
     bool CheckForDependents();
     void addDependentToSelection(App::Document* doc, App::DocumentObject* docObject);
+    void tryOfferRelabel(App::DocumentObject*& object);
```
Good catch, the direct reference was needed when this was still accessing and changing a StatusBit on the DocumentObject. I've removed &.

ashimabu commented on src/Gui/Tree.h:
```diff
@@ -257,6 +258,7 @@ private Q_SLOTS:
     static std::unique_ptr<QPixmap> documentPartialPixmap;
     std::unordered_map<const Gui::Document*,DocumentItem*> DocumentMap;
     std::unordered_map<App::DocumentObject*,std::set<DocumentObjectDataPtr> > ObjectTable;
+    std::vector<App::DocumentObject*> RelabelQueue;
```
The RelabelQueue is a container which tracks the newest Tree item. It's not truly a first in first out queue; it just tracks which objects were changed between tree operations. It actually works well as a std::unordered_set which I've changed it to.

There are two situations in TreeWidget::onUpdateStatus() that we need to handle.

1) An object with children (eg. PartDesign part) is added. onUpdateStatus() is called twice: once from slotNewObject() and once from recompute(). The relabel must be setup during the final call to onUpdateStatus() (or else we lose the cursor). The RelabelQueue is used to store the object from onUpdateStatus() #1 the object is then pulled from the queue in onUpdateStatus() #2.

2) An object without children (eg. MeshCube) is added. onUpdateStatus() is only called once and we just relabel the object directly from that call (skipping the RelabelQueue).

kadet1090 replied:
Ok, unordered set makes here perfect sense. I wonder if we can find a better name for that variable, but I guess as it is quite short-lived it is fine enough. Let's not block it over imperfect name - if anything better comes to mind we can always change it in follow-up PR, but if you want to do it now it would be great too. Thanks for explanation!

greg19 replied:
I'd hardly prefer `std::vector` from performance point of view. This collection is probably going to be small (like, 10 items at most? 20?), so there is no benefit of hash table. The unordered_set is a really sophisticated data structure that has to take care of so many things which aren't necessary here. A simple vector would be faster and use less memory. I think just coming up with a better name would be enough to get rid of the confusion. Maybe `RelabelObjects`?

kadet1090 replied:
Question here is if the performance in that case is critical enough to sacrifice semantical meaning. Set is. Structure designed to checking if something is included in set (not sure if it must be unordered here) - so the exact thing that is done here. It will be small and not used that often so there is little reason to consider just performance - premature optimization is not worth IMHO.

greg19 replied:
I don't see this as premature optimization, but using the right tool for the job. The job is to check if something is included in a small collection, and a vector is the right tool for that. If the job was to check if something is included in a big collection, an unordered set would be a right tool for it. If the job was to check if something is included in a big collection, and sometimes iterate over it in a sorted order, a set would be a right tool for the job. (If it has to be a set, please, really, don't make it a tree based one, an extra malloc is *really* unnecessary here)

The meaning can be fully captured by the variable name, or a comment if *really* necessary. There is no reason to unnecessarily slow down the software. If it can be quick, why deliberately choose to make it slow?

There is a reason why parts of FreeCAD are written in C++ and not in Python. Sometimes developers use less convenient things because they care about user experience. Please, take this into account and don't unnecessarily bloat the app.

If I understand the code correctly, the inserts will be ran regardless if the user actually subscribed to this feature – so one could even argue it's a breaking change.

kadet1090 commented:
I have tested the PR and it does not seem to work as expected. With the option enabled I'm still not offered with ability to put the name after creating the object. It seems to work only if I create new body via create sketch.

kadet1090 commented:
No, I had not attached debugger in that case. Only later I did to see if I can quickly find the reason for why it is not working but I was not able to figure out anything useful. The method `tryOfferRelabel` was called two times per every object created. first time it fails on first condition, second time it fails on second condition.

hyarion commented on src/Gui/Tree.cpp:
```diff
@@ -3061,9 +3066,17 @@ void TreeWidget::onUpdateStatus()
             if (docItem->ObjectMap.count(obj))
                 continue;
-            auto vpd = Base::freecad_dynamic_cast<ViewProviderDocumentObject>(gdoc->getViewProvider(obj));
+            auto vpd =
+                Base::freecad_dynamic_cast<ViewProviderDocumentObject>(gdoc->getViewProvider(obj)); 
```
`Base::freecad_dynamic_cast<T>()` has been replaced with `freecad_cast<T*>()`:
* https://github.com/FreeCAD/FreeCAD/pull/20515
* https://github.com/FreeCAD/FreeCAD/pull/20694

I think this might be the reason why this branch has a collision, fixing it would look like:

```suggestion
            auto vpd = freecad_cast<ViewProviderDocumentObject*>(gdoc->getViewProvider(obj)); 
```

chennes commented:
I just tested this on Windows, and it only seems to work for the first body: each subsequent body just creates the body as it used to, without activating the label for editing.

ashimabu commented:
Thanks @kadet1090 @chennes for taking the time to build and test this. I will work on a fix for this. In the meanwhile, I'll mark this PR as a draft.

maxwxyz commented:
@ashimabu do you plan on finishing this PR?

ashimabu commented:
> @ashimabu do you plan on finishing this PR?

I do have local changes to push. Still a few improvements to make, however. If I can't finish them in the next few weeks, I will close this PR in the interest of decluttering.

ashimabu commented:
The most recent push fixes the issues brought up previously. 

Regarding, #21457 this branch universally focuses the newest tree object. It also contains a user preference to automatically put new object tree labels into edit mode. This option is off by default.

New behavior with user preference on
![NewBehaviorOn](https://github.com/user-attachments/assets/5a3c5752-6e0a-4202-a15c-612bf07b33fb)

New behavior with user preference off
![NewBehaviorOff](https://github.com/user-attachments/assets/f425fd7e-0a13-4760-81a2-1e766d9dc80c)

Old behavior
![OldBehavior](https://github.com/user-attachments/assets/d77b9206-af47-4e3f-867a-36bc54d601fe)

kadet1090 commented:
Hi! This PR has seemingly a lot of conflicts due to recent update to the code style. Unfortunately in case of that PR formatting changes are not the only conflicts so I couldn't update it automatically. Instruction on how to best deal with such massive automatic changes to the codebase is here: https://github.com/FreeCAD/FreeCAD/discussions/25217. 

While it would be awesome for you to do all the steps necessary, if you resolve only conflicts before changes that is fine too - I can take care of formatting changes. In case of any troubles, feel free to ping me here (or on any other platform) and I'll try to help.
